nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -326,6 +378,108 @@ impl DataStore {
+    /// `saga_id` at the current generation.
+    ///
+    /// # Arguments
+    ///
+    /// - `instance_id`: the UUID of the Instance to lock
+    /// - `current_gen`: the current generation of the instance's `updater_id`
+    /// - `saga_id`: the UUID of the saga that's attempting to lock this
+    ///   instance.
+    ///
+    /// # Returns
+    ///
+    /// - [`Ok`]`(`[`Some`]`(`[`InstanceUpdaterLock`]`))` if the lock was acquired.
+    /// - [`Ok`]`(`[`None`]`)` if the lock could not be acquired (i.e. the
+    ///   generation has advanced since `current_gen` was captured)
+    /// - [`Err`] if a database error occurred.
+    pub async fn instance_updater_try_lock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        current_gen: Generation,
+        saga_lock_id: &Uuid,
+    ) -> Result<Option<Generation>, Error> {
+        use db::schema::instance::dsl;
+
+        // The generation to advance to.
+        let new_gen = Generation(current_gen.0.next());
+
+        let instance_id = authz_instance.id();
+
+        let locked = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            // If the generation is the same as the captured generation, we can
+            // lock this instance.
+            .filter(dsl::updater_gen.eq(current_gen))
+            .set((
+                dsl::updater_gen.eq(new_gen),
+                dsl::updater_id.eq(Some(*saga_lock_id)),
+            ))
+            .check_if_exists::<Instance>(instance_id)
+            .execute_and_check(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map(|r| match r.status {
+                UpdateStatus::Updated => Some(new_gen),
+                UpdateStatus::NotUpdatedButExists => None,
+            })
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Instance,
+                        LookupType::ById(instance_id),
+                    ),
+                )
+            })?;
+
+        Ok(locked)
+    }
+
+    /// Release the instance-updater lock.
+    pub async fn instance_updater_unlock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        saga_lock_id: &Uuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
+
+        let instance_id = authz_instance.id();
+
+        let unlocked = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(authz_instance.id()))
+            // Only unlock the instance if the provided updater ID matches that
+            // of the saga that has currently locked this instance.
+            .filter(dsl::updater_id.eq(Some(*saga_lock_id)))
+            .set((
+                dsl::updater_gen.eq(dsl::updater_gen + 1),
```

Comment:
Should we be extra cautious, and only perform this "unlock" operation is the `update_gen` is what we expect it to be?

I know we don't have a great mechanism for "saga handoff between nexuses" yet, but if we ever wanted to do that, this seems like it would be an important safety-check.

Imagine:

- Nexus A starts this saga (locks, gets partway through work)
- Nexus A becomes unresponsive, saga hangs
- Nexus B wants to take over saga (mechanism: ???)
- Nexus A somehow kicks itself back awake (yikes!), tries contacting the instance (hopefully we have [some form of fencing](https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html) to prevent the sled agent from acting on that request), and it **should** fail. However, it'll inevitably try to unlock the instance, maybe by unwinding the saga.

Reply:
Ah, yeah, because the saga UUID might end up under the ownership of another Nexus instance, that's a good point.